File: scripts/bindings/ggplot_binned_numeric.py
"""FIX-GG04: binned numeric mapping and primary guide-build failures."""
import json,math,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
def descriptor(case):
 kind=case['palette'];control=case['control'];area=kind=='area'
 breaks={'Explicit':[1,3,5,7,9]} if control=='explicit' else {'Explicit':[]} if control in ('empty_breaks','null_breaks') else {'Equal':8 if control=='count_eight' else 16} if control in ('count_eight','count_sixteen') else {'Nice':5}
 limits=[-1,5] if control=='limits' else [None,5] if control=='partial_limits' else None
 return {'function':{'Interpolated':{'normalization':{'Ggplot':{'family':'Linear','domain':[0,1],'reverse':False,'rescaler':'Maximum' if area else 'Range'}},'output':{'Interpolate':{'operation':'PowerRange','range':([0,{'custom_range':9,'reverse_range':2,'constant_range':0}[control]] if area else {'custom_range':[2,9],'reverse_range':[9,2],'constant_range':[3.5,3.5]}[control]) if control in ('custom_range','reverse_range','constant_range') else [0,6] if area else [0.1,1] if kind=='alpha' else [1,6],'exponent':0.5 if kind in ('size','area') else 1,'absolute':area}},'unknown':{'kind':'Missing'}}},'training':'Eligible','ggplot':{'Binned':{'empty_population':False,'nonfinite_population':False,'limits':limits,'oob':'Squish','breaks':breaks,'right':control!='left'}},'guide':('Hidden' if control=='null_breaks' else {'BinnedBins':'Automatic'}) if constructor_guides else {'Binned':'Automatic'}}
constructor_guides="--constructor-guides" in sys.argv
records=[]
for index,case in enumerate(json.loads((ROOT/'fixtures/parity/ggplot2/binned-numeric-palettes.json').read_text())['cases']):
 if case['control']=='left' and case['palette'] in ('size','linewidth'):continue
 owned=[];expected=case['result'];failure=expected.get('error')
 try:
  inputs=case['inputs'];kind=case['palette'];channel='StrokeWidth' if kind=='linewidth' else 'Alpha' if kind=='alpha' else 'Size'
  data=c.Data.columns({'x':c.column(list(map(float,range(len(inputs)))),kind='float64'),'v':c.column([v or 0 for v in inputs],kind='float64').validity([v is not None for v in inputs])});owned.append(data)
  def layer():return (c.rule() if kind=='linewidth' else c.points()).name('marks').stroke('#000000').numeric_scale(channel,'v',descriptor(case))
  plot=c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y(1.).x2('x').y2(2.)).layer(layer()).build();owned.append(plot)
  wire=plot.to_json();version=json.loads(wire)['version'];restored=c.Plot.from_json(wire);owned.append(restored);assert restored.to_json()==wire
  for state in ('original','edited'):
   current=restored if state=='original' else restored.edit().layer('marks',layer()).build()
   if current is not restored:owned.append(current)
   chart=current.chart();owned.append(chart);actual=chart.semantics()['layers'][0];assert not failure,(index,failure)
   wanted=expected['values']
   if len(wanted)==1:wanted=wanted*len(inputs)
   if kind!='alpha':wanted=[v for v in wanted if v is not None]
   styles=actual.get('styles',[]);assert len(styles)==len(wanted),(index,styles,wanted)
   for style,want in zip(styles,wanted):
    if kind=='alpha':assert style['color']['alpha']==(255 if want is None else round(max(0,min(1,want))*255)),(index,style,want)
    else:assert abs(style['stroke_width' if kind=='linewidth' else 'radius']-want)<2e-12,(index,style,want)
   record={'index':index,'state':state,'version':version,'styles':styles}
   if constructor_guides:
    entries=[e for group in actual.get('numeric_value_guides',{}).values() for e in group if e['visible']]
    def number(v):return None if v['kind'] in ('Missing','Null') else v['value'].get('number') if isinstance(v['value'],dict) else v['value']
    keys=[{'values':[i/(len(entries)-1) for i in range(len(entries))],'labels':[e.get('label') for e in entries],'mapped':[number(e['mapped']) for e in entries]}] if entries else []
    wanted_keys=expected['guide_keys'];assert len(keys)==len(wanted_keys),(index,keys,wanted_keys)
    for left,right in zip(keys,wanted_keys):
     assert left['labels']==right['labels'],(index,left,right)
     for field in ('values','mapped'):
      assert len(left[field])==len(right[field]),(index,left,right)
      for a,b in zip(left[field],right[field]):assert (a is None and b is None) or (a is not None and b is not None and abs(a-b)<2e-12),(index,field,a,b)
    record['guide_keys']=keys
   records.append(record)
   if case['population']=='many' and case['control']=='count_eight' and state=='original':
    request=output.request(current,options);owned.append(request);frame=request.prepare();owned.append(frame)
    for fmt in ('svg','pdf','png'):(out/f'numeric-{kind}.{fmt}').write_bytes(frame.export(fmt))
 except c.ChartError as error:
  diagnostic=json.loads(str(error));assert failure,(index,diagnostic);assert diagnostic['code'] in ('CHART_NUMERICAL_DOMAIN','CHART_VALIDATION'),(index,diagnostic)
  records.append({'index':index,'error':diagnostic['code']})
 finally:
  for value in reversed(owned):value.dispose()
assert len(records)==502,len(records)
(out/'binned-numeric-records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose();print('PASS Python: 502 binned numeric states; twelve publication files.')

File: crates/chart-core/tests/ggplot_binned_numeric.rs
//! FIX-GG04: built-in numeric palettes composed with the shared binned policy.
use chart_core::{
    interpolate::{Number, Value},
    scales::*,
};
#[test]
fn binned_numeric_maps_match_reference() {
    let fixture: serde_json::Value = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/binned-numeric-palettes.json"
    ))
    .unwrap();
    let cases = fixture["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 288);
    let mut compared = 0;
    for case in cases {
        // R size/linewidth wrappers have no `right` argument. Our explicit generic
        // binned policy corresponds to binned_scale, which does accept it.
        if case["control"] == "left"
            && (case["palette"] == "size" || case["palette"] == "linewidth")
        {
            continue;
        }
        compared += 1;
        let kind = match case["palette"].as_str().unwrap() {
            "size" => GgplotNumericPalette::Size,
            "area" => GgplotNumericPalette::Area,
            "alpha" => GgplotNumericPalette::Alpha,
            "linewidth" => GgplotNumericPalette::Linewidth,
            _ => unreachable!(),
        };
        let control = case["control"].as_str().unwrap();
        let policy = GgplotBinnedPolicy {
            breaks: match control {
                "explicit" => GgplotBreaks::Explicit([1., 3., 5., 7., 9.].map(Number).to_vec()),
                "empty_breaks" | "null_breaks" => GgplotBreaks::Explicit(vec![]),
                "count_eight" => GgplotBreaks::Equal(8.),
                "count_sixteen" => GgplotBreaks::Equal(16.),
                _ => GgplotBreaks::Nice(5.),
            },
            limits: match control {
                "limits" => Some([Some(Number(-1.)), Some(Number(5.))]),
                "partial_limits" => Some([None, Some(Number(5.))]),
                _ => None,
            },
            right: control != "left",
            ..Default::default()
        };
        let inputs: Vec<_> = case["inputs"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_f64().map(Number))
            .collect();
        check_constructor(case, kind, policy.clone(), &inputs);
        let result = (|| -> chart_core::ChartResult<_> {
            let spec = numeric_spec(case, kind)?
                .with_ggplot(GgplotScalePolicy::Binned(Box::new(policy)))?
                .with_guide(GgplotScaleGuide::Binned(GgplotGuideLabels::Automatic))?
                .trained(&inputs)?;
            let json = serde_json::to_string(&spec).unwrap();
            let scale = MappedScale::for_numbers(serde_json::from_str(&json).unwrap())?;
            let values = inputs
                .iter()
                .map(|input| scale.numeric(input.map(|v| v.0)))
                .collect::<chart_core::ChartResult<Vec<_>>>()?;
            let guide = scale.binned_guide_entries(4096, 65536)?;
            Ok((values, guide))
        })();
        let expected = &case["raw_result"];
        if expected.get("error").is_some() {
            assert!(result.is_err(), "{case}");
            continue;
        }
        let (values, guide) = result.unwrap_or_else(|error| panic!("{error:?}: {case}"));
        let wanted = expected["values"].as_array().unwrap();
        assert!(wanted.len() == values.len() || wanted.len() == 1, "{case}");
        for (i, got) in values.iter().enumerate() {
            let want = &wanted[if wanted.len() == 1 { 0 } else { i }];
            if want.is_null() {
                assert_eq!(*got, Value::Missing, "{case}");
            } else {
                let Value::Number(got) = got else {
                    panic!("{got:?} {case}")
                };
                assert!(
                    (got.0 - want.as_f64().unwrap()).abs() < 2e-12,
                    "{got:?} {case}"
                );
            }
        }
        let entries = guide.unwrap();
        let cuts = expected["guide"]["breaks"].as_array().unwrap();
        assert_eq!(entries.len(), cuts.len(), "{case}");
        for (got, want) in entries.iter().zip(cuts) {
            assert!(
                (got.transformed.0 - want.as_f64().unwrap()).abs() < 2e-12,
                "{case}"
            );
        }
        assert_eq!(
            entries
                .iter()
                .map(|e| e.label.as_deref())
                .collect::<Vec<_>>(),
            expected["guide"]["labels"]
                .as_array()
                .unwrap()
                .iter()
                .map(|v| v.as_str())
                .collect::<Vec<_>>(),
            "{case}"
        );
    }
    assert_eq!(compared, 276);
}

fn check_constructor(
    case: &serde_json::Value,
    kind: GgplotNumericPalette,
    policy: GgplotBinnedPolicy,
    inputs: &[Option<Number>],
) {
    use chart_core::prelude::*;
    let expected = &case["result"];
    let result = (|| -> chart_core::ChartResult<_> {
        let scale = numeric_spec(case, kind)?
            .with_ggplot(GgplotScalePolicy::Binned(Box::new(policy)))?
            .with_guide(if case["control"] == "null_breaks" {
                GgplotScaleGuide::Hidden
            } else {
                GgplotScaleGuide::BinnedBins(GgplotGuideLabels::Automatic)
            })?;
        let data = Data::columns()
            .column("x", (0..inputs.len()).map(|i| i as f64).collect::<Vec<_>>())
            .column(
                "v",
                inputs.iter().map(|v| v.map(|v| v.0)).collect::<Vec<_>>(),
            )
            .build()?;
        let channel = match kind {
            GgplotNumericPalette::Alpha => NumericAesthetic::Alpha,
            GgplotNumericPalette::Linewidth => NumericAesthetic::StrokeWidth,
            _ => NumericAesthetic::Size,
        };
        let layer = if kind == GgplotNumericPalette::Linewidth {
            rule()
        } else {
            points()
        };
        let plot = plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes().x("x").y(1.).x2("x").y2(2.))
            .layer(layer.numeric_scale(channel, "v", scale))
            .build()?;
        let wire = plot.to_json()?;
        let restored = Plot::from_json(&wire)?;
        assert_eq!(wire, restored.to_json()?);
        restored.chart()?.prepare()
    })();
    if expected.get("error").is_some() || expected["draw_error"].is_string() {
        assert!(result.is_err(), "{case}");
        return;
    }
    let prepared = result.unwrap_or_else(|e| panic!("{e:?}: {case}"));
    let entries = prepared.layers()[0]
        .numeric_value_guides()
        .values()
        .flatten()
        .filter(|e| e.visible)
        .collect::<Vec<_>>();
    let keys = expected["guide_keys"].as_array().unwrap();
    if entries.is_empty() {
        assert!(keys.is_empty(), "{case}");
        return;
    }
    assert_eq!(keys.len(), 1, "{case}");
    let keys = &keys[0];
    assert_eq!(
        entries.len(),
        keys["values"].as_array().unwrap().len(),
        "{case}"
    );
    for (index, entry) in entries.iter().enumerate() {
        let position = index as f64 / (entries.len() - 1) as f64;
        assert!(
            (position - keys["values"][index].as_f64().unwrap()).abs() < 2e-12,
            "{case}"
        );
        assert_eq!(
            entry.label.as_deref(),
            keys["labels"][index].as_str(),
            "{case}"
        );
        let want = &keys["mapped"][index];
        if want.is_null() {
            assert_eq!(entry.mapped, Some(Value::Missing), "{case}");
        } else {
            let Some(Value::Number(got)) = entry.mapped else {
                panic!("{case}")
            };
            assert!((got.0 - want.as_f64().unwrap()).abs() < 2e-12, "{case}");
        }
    }
}

fn numeric_spec(
    case: &serde_json::Value,
    kind: GgplotNumericPalette,
) -> chart_core::ChartResult<MappedScaleSpec> {
    let mut spec = ggplot_numeric_default(kind)?;
    let control = case["control"].as_str().unwrap();
    let range = if kind == GgplotNumericPalette::Area {
        match control {
            "custom_range" => Some([0., 9.]),
            "reverse_range" => Some([0., 2.]),
            "constant_range" => Some([0., 0.]),
            _ => None,
        }
    } else {
        match control {
            "custom_range" => Some([2., 9.]),
            "reverse_range" => Some([9., 2.]),
            "constant_range" => Some([3.5, 3.5]),
            _ => None,
        }
    };
    if let Some(range) = range {
        let ScaleFunctionSpec::Interpolated(ref mut interpolated) = spec.function else {
            unreachable!()
        };
        let ScaleRangeFunction::Interpolate(
            chart_core::interpolate::InterpolationSpec::PowerRange {
                range: ref mut output_range,
                ..
            },
        ) = interpolated.output
        else {
            unreachable!()
        };
        *output_range = range.map(Number);
    }
    Ok(spec)
}
